pkg/ccl/streamingccl/streamproducer/replication_stream_test.go

```diff
@@ -835,3 +808,218 @@ USE d;
+				MockSpanConfigTableName: dummyFQN,
+			},
+		},
+	})
+	defer cleanup()
+
+	h.SysSQL.Exec(t, `
+CREATE DATABASE d;
+USE d;`)
+	h.SysSQL.Exec(t, fmt.Sprintf("CREATE TABLE %s (LIKE system.span_configurations INCLUDING ALL)", dummyFQN))
+
+	tenantID := roachpb.MustMakeTenantID(uint64(10))
+	tenantName := roachpb.TenantName("app")
+
+	_, tenantCleanup := h.CreateTenant(t, tenantID, tenantName)
+	defer tenantCleanup()
+
+	tenantCodec := keys.MakeSQLCodec(tenantID)
+	accessor := spanconfigkvaccessor.New(
+		h.SysServer.DB(),
+		h.SysServer.InternalExecutor().(isql.Executor),
+		h.SysServer.ClusterSettings(),
+		h.SysServer.Clock(),
+		dummyFQN.String(),
+		nil, /* knobs */
+	)
+
+	const streamSpanConfigsQuery = `SELECT * FROM crdb_internal.setup_span_configs_stream($1)`
+	source, feed := startReplication(ctx, t, h, makePartitionStreamDecoder,
+		streamSpanConfigsQuery, tenantName)
+	defer feed.Close(ctx)
+
+	// Add a few more use cases:
+	// -  update a subset of the tenant key space
+
+	// makes a dummy span config record with the given ttl. If the ttl is 0,  no span config is added,
+	// and the record is treated as a delete record.
+	makeRecord := func(targetSpan roachpb.Span, ttl int) spanconfig.Record {
+		target := spanconfig.MakeTargetFromSpan(targetSpan)
+		var spanConfig roachpb.SpanConfig
+		if ttl > 0 {
+			spanConfig = roachpb.SpanConfig{
+				GCPolicy: roachpb.GCPolicy{
+					TTLSeconds: int32(ttl),
+				},
+			}
+		}
+		// check that all orderedUpdates are observed.
+		record, err := spanconfig.MakeRecord(target, spanConfig)
+		require.NoError(t, err)
+		return record
+	}
+
+	recordToEntry := func(record spanconfig.Record) roachpb.SpanConfigEntry {
+		t := record.GetTarget().ToProto()
+		c := record.GetConfig()
+		return roachpb.SpanConfigEntry{
+			Target: t,
+			Config: c,
+		}
+	}
+
+	makeTableSpan := func(highTableID uint32) roachpb.Span {
+		syntheticTableSpanPrefix := tenantCodec.TablePrefix(highTableID)
+		return roachpb.Span{Key: syntheticTableSpanPrefix, EndKey: syntheticTableSpanPrefix.PrefixEnd()}
+	}
+
+	irrelevantTenant := keys.MakeSQLCodec(roachpb.MustMakeTenantID(231))
+	relevantSpan := tenantCodec.TenantSpan()
+	t1Span, t2Span, t3Span := makeTableSpan(1001), makeTableSpan(1002), makeTableSpan(1003)
+	t123Span := roachpb.Span{Key: t1Span.Key, EndKey: t3Span.EndKey}
+	expectedSpanConfigs := makeSpanConfigUpdateRecorder()
+
+	expectedUpdateCount := 0
+	for ts, toApply := range []struct {
+		updates []spanconfig.Record
+		deletes []spanconfig.Target
+	}{
+		{
+			// This irrelevant span should not surface.
+			updates: []spanconfig.Record{makeRecord(irrelevantTenant.TenantSpan(), 3)},
+		},
+		{
+			updates: []spanconfig.Record{makeRecord(t1Span, 2)},
+		},
+		{
+			// Update the Span.
+			updates: []spanconfig.Record{makeRecord(t1Span, 3)},
+		},
+		{
+			// Add Two new records at the same time
+			updates: []spanconfig.Record{makeRecord(t2Span, 2), makeRecord(t3Span, 5)},
+		},
+		{
+			// Merge these Records
```

Comment:
@arulajmani in this test, I apply a few span config updates for dummy tables via the kvAccessor on a dummy span config table and check that they get picked up by the c2c machinery (which you don't need to worry about). 

In this test case, I merge 3 span config records for t1, t2, and t3. One surprising thing I noticed is that the a `delete` of the spanconfig record for t1 no-ops in the kvaccessor, and consequently never replicates. I think this occurs because the update on the t123 span updates that record instead. Is this behavior surprising?

I have a similar question for the Split test case below: when I split the record up, a delete on the merged record noops as well. Is this surprising?

Reply:
> One surprising thing I noticed is that the a delete of the spanconfig record for t1 no-ops in the kvaccessor, and consequently never replicates.

I'm probably misunderstanding what you're saying, but where is the delete for t1? Isn't there just a delete for t2 and t3? 

```go
deletes: []spanconfig.Target{spanconfig.MakeTargetFromSpan(t2Span), spanconfig.MakeTargetFromSpan(t3Span)},
```

Are you saying the same happens for something like as well?:

```go
deletes: []spanconfig.Target{spanconfig.MakeTargetFromSpan(t1Span), spanconfig.MakeTargetFromSpan(t2Span), spanconfig.MakeTargetFromSpan(t3Span)},
```

If so, I think you're right that it's because t123 span updates that record. Note that the primary key of `system.span_configurations` is just the start key -- so I think what you're seeing is a deletion followed by a write. Given this is all happening in a single transaction, maybe the rangefeed considers all this to be an update for `t1Span.StartKey`?

Reply:
> I imagine there's some logic above raft which, for a given transaction, no-ops all the intermediate operations for a given key. I.e. if within a txn:

Intermediate operations are only visible to the transaction itself. They're never visible to any other concurrent actor in the system. Intermediate values are stored inside of the intent, whereas the most recent value is stored in the MVCC keyspace. So when the intent is resolved, any information about intermediate operations leaves with it. I think that's what you're seeing here.